File: src/markdown.ts
import { readFileSync } from "node:fs";

export type Attributes = Record<string, unknown>;

export interface MarkdownDocument {
  attributes: Attributes;
  body: string;
}

export function parseMarkdown(contents: string): MarkdownDocument {
  const lines = contents.replaceAll("\r\n", "\n").split("\n");
  if (lines[0] !== "---") {
    throw new Error("Markdown document must start with YAML front matter");
  }
  const end = lines.indexOf("---", 1);
  if (end === -1) {
    throw new Error("Markdown document has unterminated front matter");
  }
  const attributes: Attributes = {};
  for (const line of lines.slice(1, end)) {
    if (line.trim() === "" || line.trimStart().startsWith("#")) continue;
    const separator = line.indexOf(":");
    if (separator === -1) throw new Error(`Invalid front matter line: ${line}`);
    const key = line.slice(0, separator).trim();
    attributes[key] = parseValue(line.slice(separator + 1).trim());
  }
  return {
    attributes,
    body: lines
      .slice(end + 1)
      .join("\n")
      .replace(/^\n/, ""),
  };
}

export function readMarkdown(path: string): MarkdownDocument {
  return parseMarkdown(readFileSync(path, "utf8"));
}

export function renderMarkdown(document: MarkdownDocument): string {
  const frontMatter = Object.entries(document.attributes)
    .map(([key, value]) => `${key}: ${renderValue(value)}`)
    .join("\n");
  return `---\n${frontMatter}\n---\n\n${document.body.trim()}\n`;
}

export function stringAttribute(attributes: Attributes, key: string): string {
  const value = attributes[key];
  if (typeof value !== "string" || value.length === 0) {
    throw new Error(`Expected non-empty string attribute: ${key}`);
  }
  return value;
}

export function stringsAttribute(attributes: Attributes, key: string): string[] {
  const value = attributes[key];
  if (!Array.isArray(value) || value.some((item) => typeof item !== "string")) {
    throw new Error(`Expected string array attribute: ${key}`);
  }
  return value as string[];
}

export function recordAttribute(attributes: Attributes, key: string): Record<string, string> {
  const value = attributes[key];
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`Expected object attribute: ${key}`);
  }
  if (Object.values(value).some((item) => typeof item !== "string")) {
    throw new Error(`Expected string values in attribute: ${key}`);
  }
  return value as Record<string, string>;
}

function parseValue(value: string): unknown {
  if (value === "") return "";
  if (
    value.startsWith('"') ||
    value.startsWith("[") ||
    value.startsWith("{") ||
    value === "true" ||
    value === "false" ||
    value === "null" ||
    /^-?\d+(\.\d+)?$/.test(value)
  ) {
    try {
      return JSON.parse(value);
    } catch {
      throw new Error(`Front matter value is not valid JSON-compatible YAML: ${value}`);
    }
  }
  return value;
}

function renderValue(value: unknown): string {
  if (typeof value === "string") return JSON.stringify(value);
  return JSON.stringify(value);
}

File: src/agora/markdown.py
import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any

Attributes = dict[str, Any]


@dataclass
class MarkdownDocument:
    attributes: Attributes
    body: str


def parse_markdown(contents: str) -> MarkdownDocument:
    lines = contents.replace("\r\n", "\n").split("\n")
    if not lines or lines[0] != "---":
        raise ValueError("Markdown document must start with YAML front matter")
    try:
        end = lines.index("---", 1)
    except ValueError as error:
        raise ValueError("Markdown document has unterminated front matter") from error

    attributes: Attributes = {}
    for line in lines[1:end]:
        if not line.strip() or line.lstrip().startswith("#"):
            continue
        if ":" not in line:
            raise ValueError(f"Invalid front matter line: {line}")
        key, value = line.split(":", 1)
        attributes[key.strip()] = _parse_value(value.strip())

    body = "\n".join(lines[end + 1 :])
    if body.startswith("\n"):
        body = body[1:]
    return MarkdownDocument(attributes=attributes, body=body)


def read_markdown(path: Path) -> MarkdownDocument:
    return parse_markdown(path.read_text(encoding="utf-8"))


def render_markdown(document: MarkdownDocument) -> str:
    front_matter = "\n".join(
        f"{key}: {_render_value(value)}" for key, value in document.attributes.items()
    )
    return f"---\n{front_matter}\n---\n\n{document.body.strip()}\n"


def string_attribute(attributes: Attributes, key: str) -> str:
    value = attributes.get(key)
    if not isinstance(value, str) or not value:
        raise ValueError(f"Expected non-empty string attribute: {key}")
    return value


def strings_attribute(attributes: Attributes, key: str) -> list[str]:
    value = attributes.get(key)
    if not isinstance(value, list) or any(not isinstance(item, str) for item in value):
        raise ValueError(f"Expected string array attribute: {key}")
    return value


def record_attribute(attributes: Attributes, key: str) -> dict[str, str]:
    value = attributes.get(key)
    if not isinstance(value, dict) or any(
        not isinstance(name, str) or not isinstance(item, str) for name, item in value.items()
    ):
        raise ValueError(f"Expected string map attribute: {key}")
    return value


def _parse_value(value: str) -> Any:
    if not value:
        return ""
    if value[0] in {'"', "[", "{"} or value in {"true", "false", "null"}:
        try:
            return json.loads(value)
        except json.JSONDecodeError as error:
            raise ValueError(
                f"Front matter value is not valid JSON-compatible YAML: {value}"
            ) from error
    return value


def _render_value(value: Any) -> str:
    return json.dumps(value, ensure_ascii=True, separators=(",", ":"))
